folding-schemes/src/commitment/ipa.rs

```diff
@@ -0,0 +1,679 @@
+
+    <C as ark_ec::CurveGroup>::BaseField: ark_ff::PrimeField,
+    for<'a> &'a GC: GroupOpsBounds<'a, C, GC>,
+{
+    /// verify the IPA opening proof, K=log2(d), where d is the degree of the committed polynomial,
+    /// and H indicates if the commitment is in hiding mode and thus uses blinding factors, if
+    /// not, there are some constraints saved.
+    #[allow(clippy::too_many_arguments)]
+    pub fn verify<const K: usize>(
+        g: &Vec<GC>,                                  // parms.generators
+        h: &GC,                                       // parms.h
+        x: &NonNativeFieldVar<C::ScalarField, CF<C>>, // evaluation point
+        v: &NonNativeFieldVar<C::ScalarField, CF<C>>, // value at evaluation point
+        P: &GC,                                       // commitment
+        p: &ProofVar<C, GC>,
+        r: &NonNativeFieldVar<C::ScalarField, CF<C>>, // blinding factor
+        u: &[NonNativeFieldVar<C::ScalarField, CF<C>>; K], // challenges
+        U: &GC,                                       // challenge
+    ) -> Result<Boolean<CF<C>>, SynthesisError> {
+        if p.L.len() != K || p.R.len() != K {
+            return Err(SynthesisError::Unsatisfiable);
+        }
+
+        let P_ = P + U.scalar_mul_le(v.to_bits_le()?.iter())?;
+        let mut q_0 = P_;
+        let mut r = r.clone();
+
+        // compute u[i]^-1 once
+        let mut u_invs = vec![NonNativeFieldVar::<C::ScalarField, CF<C>>::zero(); u.len()];
+        for (j, u_j) in u.iter().enumerate() {
+            u_invs[j] = u_j.inverse()?;
+        }
+
+        // compute b & G from s
+        // let d: usize = 2_u64.pow(K as u32) as usize;
+        let s = build_s_gadget(u, &u_invs, K)?;
+        // b = <s, b_vec> = <s, [1, x, x^2, ..., x^K-1]>
+        let b = s_b_inner_gadget(u, x)?;
+        // ensure that generators.len() === s.len():
+        if g.len() < K {
+            return Err(SynthesisError::Unsatisfiable);
+        }
+
+        // msm: G=<G, s>
+        let mut G = GC::zero();
+        for (i, s_i) in s.iter().enumerate() {
+            G += g[i].scalar_mul_le(s_i.to_bits_le()?.iter())?;
+        }
+
+        for (j, u_j) in u.iter().enumerate() {
+            let uj2 = u_j.square()?;
+            let uj_inv2 = u_invs[j].square()?; // cheaper square than inversing the uj2
+
+            q_0 = q_0
+                + p.L[j].scalar_mul_le(uj2.to_bits_le()?.iter())?
+                + p.R[j].scalar_mul_le(uj_inv2.to_bits_le()?.iter())?;
+            if H {
+                r = r + &p.l[j] * &uj2 + &p.r[j] * &uj_inv2;
+            }
+        }
+
+        let q_1 = if H {
+            G.scalar_mul_le(p.a.to_bits_le()?.iter())?
+                + h.scalar_mul_le(r.to_bits_le()?.iter())?
+                + U.scalar_mul_le((p.a.clone() * b).to_bits_le()?.iter())?
+        } else {
+            G.scalar_mul_le(p.a.to_bits_le()?.iter())?
+                + U.scalar_mul_le((p.a.clone() * b).to_bits_le()?.iter())?
+        };
```

Comment:
It is intended, yes. The idea is that `H` is a constant that is defined when instantiating the folding scheme, because the circuit is different (and cheaper) if it does not use the hiding than if it uses it. As for example, in our initial use case, we won't use blinding factors (neither used in microsoft/Nova), so we can save constraints in our circuits.
Regarding ProverKey, it was already not reusable for example in the decider circuit (where we use these commitments gadgets) since the circuit has the r1cs matrices hardcoded.